libs/ledger/src/protocols/notarisation_service.cpp

```diff
@@ -0,0 +1,392 @@
+                                              previous_height_notarisations.end())
+        {
+          notarisations_built_[next_height][hash] = sig;
+        }
+        else
+        {
+          // If previous block is not notarised save for later
+          // TODO(JMW): Processing of detached notarised blocks
+          detached_notarisations_built_[next_height][hash] = sig;
+        }
+      }
+      else
+      {
+        // If do not the block then save for processing later
+        detached_notarisations_built_[next_height][hash] = sig;
+      }
+    }
+
+    // If we have obtained one linked notarisation for this block height then continue to next one
+    if (!notarisations_built_[next_height].empty())
+    {
+      return State::COMPLETE;
+    }
+  }
+
+  return State::COLLECT_NOTARISATIONS;
+}
+
+NotarisationService::State NotarisationService::OnComplete()
+{
+  FETCH_LOCK(mutex_);
+  uint64_t next_height = NextBlockHeight();
+  for (const auto &new_notarisation : notarisations_built_[next_height])
+  {
+    callback_(new_notarisation.first);
+  }
+  highest_notarised_block_height_ = next_height;
+
+  // TODO(JMW): Clear old signature shares
+
+  // Completed notarisation of a sequence of blocks during aeon. Any notarised blocks not received
+  // through RPC will be obtained via broadcast
+  if (NextBlockHeight() > active_exe_unit_->aeon.round_end)
+  {
+    return State::KEY_ROTATION;
+  }
+
+  return State::COLLECT_NOTARISATIONS;
+}
+
+NotarisationService::BlockNotarisationShares NotarisationService::GetNotarisations(
+    BlockHeight const &height)
+{
+  FETCH_LOCK(mutex_);
+  if (notarisations_being_built_.find(height) == notarisations_being_built_.end())
+  {
+    return {};
+  }
+  return notarisations_being_built_.at(height);
+}
+
+void NotarisationService::NotariseBlock(BlockBody const &block)
+{
+  FETCH_LOCK(mutex_);
+
+  // Return if not eligible to notarise, block is too far in the past of the highest notarised block
+  // or beyond the window of this aeon
+  if (!active_exe_unit_ || block.block_number < active_exe_unit_->aeon.round_start ||
+      block.block_number >= active_exe_unit_->aeon.round_end)
+  {
+    return;
+  }
+  // Return if block has already been notarised
+  if (notarisations_built_.find(block.block_number) != notarisations_built_.end() &&
+      notarisations_built_.at(block.block_number).find(block.hash) !=
+          notarisations_built_.at(block.block_number).end())
+  {
+    // TODO(JMW): Block has already been notarised -> tell main chain
+    return;
+  }
+  // Return if block is too far in the past of the head of the notarised chain
+  if (block.block_number < BlockNumberCutoff())
+  {
+    return;
+  }
+
+  // If first block of the aeon then qual should match
+  if (block.block_number == active_exe_unit_->aeon.round_start)
+  {
+    assert(block.block_entropy.qualified == active_exe_unit_->manager.qual());
+  }
+
+  // Determine rank of miner in qual
```

Comment:
Since this logic also happens in consensus.cpp, it might be worth thinking about making the assertion that the BC will only update the notarisation service with ever heavier blocks